File: library/src/main/java/com/dino/lib/dialog/list/PriorityListDialog.java
package com.dino.lib.dialog.list;

import android.content.Context;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.text.TextUtils;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.TextView;
import com.dino.lib.dialog.BaseDialogBuilder;
import com.dino.lib.dialog.BaseDialogFragment;
import com.dino.lib.dialog.R;
import java.util.Arrays;

/**
 * Created by dineshkumar.m on 11/03/18.
 */

public class PriorityListDialog extends BaseDialogFragment {

    private CharSequence mTitle;
    private CharSequence[] mItems;
    @ChoiceMode
    private int mMode;
    private int[] mCheckedItems = new int[]{};
    private CharSequence mCancelButtonText;
    private CharSequence mConfirmButtonText;
    private String mTag;
    private ListDialogListener mListDialogListener;
    private MultiChoiceListDialogListener mMultiChoiceListDialogListener;

    public static SimpleListDialogBuilder createBuilder(Context context, FragmentManager fragmentManager) {
        return new SimpleListDialogBuilder(context, fragmentManager);
    }

    private static int[] asIntArray(SparseBooleanArray checkedItems) {
        int checked = 0;
        // compute number of items
        for (int i = 0; i < checkedItems.size(); i++) {
            int key = checkedItems.keyAt(i);
            if (checkedItems.get(key)) {
                ++checked;
            }
        }

        int[] array = new int[checked];
        //add indexes that are checked
        for (int i = 0, j = 0; i < checkedItems.size(); i++) {
            int key = checkedItems.keyAt(i);
            if (checkedItems.get(key)) {
                array[j++] = key;
            }
        }
        Arrays.sort(array);
        return array;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (getArguments() == null) {
            throw new IllegalArgumentException(
                    "use SimpleListDialogBuilder to construct this dialog");
        }
    }

    private ListAdapter prepareAdapter(final int itemLayoutId) {
        return new ArrayAdapter<Object>(getActivity(),
                itemLayoutId,
                R.id.pdl_text,
                getItems()) {

            /**
             * Overriding default implementation because it ignores current light/dark theme.
             */
            @Override
            public View getView(int position, View convertView, ViewGroup parent) {
                if (convertView == null) {
                    convertView = LayoutInflater.from(parent.getContext()).inflate(itemLayoutId, parent, false);
                }
                TextView t = (TextView)convertView.findViewById(R.id.pdl_text);
                if (t != null) {
                    t.setText((CharSequence)getItem(position));
                }
                return convertView;
            }
        };
    }

    private void buildMultiChoice(Builder builder) {
        builder.setItems(
                prepareAdapter(R.layout.view_list_item_multichoice),
                getCheckedItems(), AbsListView.CHOICE_MODE_MULTIPLE,
                new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                        SparseBooleanArray checkedPositions = ((ListView)parent).getCheckedItemPositions();
                        setCheckedItems(asIntArray(checkedPositions));
                    }
                });
    }

    private void buildSingleChoice(Builder builder) {
        builder.setItems(
                prepareAdapter(R.layout.view_list_item_singlechoice),
                getCheckedItems(),
                AbsListView.CHOICE_MODE_SINGLE, new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                        SparseBooleanArray checkedPositions = ((ListView)parent).getCheckedItemPositions();
                        setCheckedItems(asIntArray(checkedPositions));
                    }
                });
    }

    private void buildNormalChoice(Builder builder) {
        builder.setItems(
                prepareAdapter(R.layout.view_list_item), -1,
                new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                        /*for (IListDialogListener listener : getSingleDialogListeners()) {
                            listener.onListItemSelected(getItems()[position], position, mRequestCode);
                        }*/
                        dismiss();
                        if(mListDialogListener != null){
                            mListDialogListener.onListItemSelected(getItems()[position], position, mRequestCode);
                        }
                    }
                });
    }

    @Override
    protected Builder build(Builder builder) {
        final CharSequence title = getTitle();
        if (!TextUtils.isEmpty(title)) {
            builder.setTitle(title);
        }

        if (!TextUtils.isEmpty(getNegativeButtonText())) {
            builder.setNegativeButton(getNegativeButtonText(), new View.OnClickListener() {
                @Override
                public void onClick(View view) {

                    dismiss();
                }
            });
        }

        if (getMode() != AbsListView.CHOICE_MODE_NONE) {
            View.OnClickListener positiveButtonClickListener = null;
            switch (getMode()) {
                case AbsListView.CHOICE_MODE_MULTIPLE:
                    positiveButtonClickListener = new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            final int[] checkedPositions = getCheckedItems();
                            final CharSequence[] items = getItems();
                            final CharSequence[] checkedValues = new CharSequence[checkedPositions.length];
                            int i = 0;
                            for (int checkedPosition : checkedPositions) {
                                if (checkedPosition >= 0 && checkedPosition < items.length) {
                                    checkedValues[i++] = items[checkedPosition];
                                }
                            }
                            dismiss();
                            if(mMultiChoiceListDialogListener != null){
                                mMultiChoiceListDialogListener.onListItemsSelected(checkedValues, checkedPositions, mRequestCode);
                            }
                        }
                    };
                    break;
                case AbsListView.CHOICE_MODE_SINGLE:
                    positiveButtonClickListener = new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            int selectedPosition = -1;
                            final int[] checkedPositions = getCheckedItems();
                            final CharSequence[] items = getItems();
                            for (int i : checkedPositions) {
                                if (i >= 0 && i < items.length) {
                                    //1st valid value
                                    selectedPosition = i;
                                    break;
                                }
                            }

                            dismiss();
                            if (selectedPosition != -1) {
                                if(mListDialogListener != null){
                                    mListDialogListener.onListItemSelected(items[selectedPosition], selectedPosition, mRequestCode);
                                }
                            } else {
                                //for (ISimpleDialogCancelListener listener : getCancelListeners()) {
                                //    listener.onCancelled(mRequestCode);
                                //}
                            }

                        }
                    };
                    break;
            }

            CharSequence positiveButton = getPositiveButtonText();
            if (TextUtils.isEmpty(getPositiveButtonText())) {
                //we always need confirm button when CHOICE_MODE_SINGLE or CHOICE_MODE_MULTIPLE
                positiveButton = getString(android.R.string.ok);
            }
            builder.setPositiveButton(positiveButton, positiveButtonClickListener);
        }

        // prepare list and its item click listener
        final CharSequence[] items = getItems();
        if (items != null && items.length > 0) {
            @ChoiceMode
            final int mode = getMode();
            switch (mode) {
                case AbsListView.CHOICE_MODE_MULTIPLE:
                    buildMultiChoice(builder);
                    break;
                case AbsListView.CHOICE_MODE_SINGLE:
                    buildSingleChoice(builder);
                    break;
                case AbsListView.CHOICE_MODE_NONE:
                    buildNormalChoice(builder);
                    break;
            }
        }

        return builder;
    }

    @Override
    protected String getCustomTag() {
        return null;
    }

    @Override
    protected int getPriority() {
        return 0;
    }

    private CharSequence getTitle() {
        return mTitle;
    }

    @SuppressWarnings("ResourceType")
    @ChoiceMode
    private int getMode() {
        return mMode;
    }

    private CharSequence[] getItems() {
        return mItems;
    }

    @NonNull
    private int[] getCheckedItems() {
        return mCheckedItems;
    }

    private void setCheckedItems(int[] checkedItems) {
        mCheckedItems = checkedItems;
    }

    private CharSequence getPositiveButtonText() {
        return mConfirmButtonText;
    }

    private CharSequence getNegativeButtonText() {
        return mCancelButtonText;
    }

    @IntDef({AbsListView.CHOICE_MODE_MULTIPLE, AbsListView.CHOICE_MODE_SINGLE, AbsListView.CHOICE_MODE_NONE})
    public @interface ChoiceMode {
    }

    public static class SimpleListDialogBuilder extends BaseDialogBuilder<SimpleListDialogBuilder> {

        private CharSequence title;

        private CharSequence[] items;

        @ChoiceMode
        private int mode;
        private int[] checkedItems;

        private CharSequence cancelButtonText;
        private CharSequence confirmButtonText;
        private ListDialogListener iListDialogListener;
        private MultiChoiceListDialogListener iMultiChoiceListDialogListener;


        public SimpleListDialogBuilder(Context context, FragmentManager fragmentManager) {
            super(context, fragmentManager, PriorityListDialog.class);
        }

        @Override
        protected SimpleListDialogBuilder self() {
            return this;
        }

        private Resources getResources() {
            return mContext.getResources();
        }

        public SimpleListDialogBuilder setTitle(CharSequence title) {
            this.title = title;
            return this;
        }

        public SimpleListDialogBuilder setTitle(int titleResID) {
            this.title = getResources().getString(titleResID);
            return this;
        }

        public SimpleListDialogBuilder setCheckedItems(int[] positions) {
            this.checkedItems = positions;
            return this;
        }


        public SimpleListDialogBuilder setSelectedItem(int position) {
            this.checkedItems = new int[]{position};
            return this;
        }

        public SimpleListDialogBuilder setChoiceMode(@ChoiceMode int choiceMode) {
            this.mode = choiceMode;
            return this;
        }

        public SimpleListDialogBuilder setItems(CharSequence[] items) {
            this.items = items;
            return this;
        }

        public SimpleListDialogBuilder setItems(int itemsArrayResID) {
            this.items = getResources().getStringArray(itemsArrayResID);
            return this;
        }

        public SimpleListDialogBuilder setConfirmButtonText(CharSequence text) {
            this.confirmButtonText = text;
            return this;
        }

        public SimpleListDialogBuilder setConfirmButtonText(int confirmBttTextResID) {
            this.confirmButtonText = getResources().getString(confirmBttTextResID);
            return this;
        }

        public SimpleListDialogBuilder setCancelButtonText(CharSequence text) {
            this.cancelButtonText = text;
            return this;
        }

        public SimpleListDialogBuilder setCancelButtonText(int cancelBttTextResID) {
            this.cancelButtonText = getResources().getString(cancelBttTextResID);
            return this;
        }

        public SimpleListDialogBuilder setTag(String tag) {
            this.mTag = tag;
            return this;
        }

        public SimpleListDialogBuilder setOnListItemSelected(ListDialogListener listener) {
            this.iListDialogListener = listener;
            return this;
        }
        public SimpleListDialogBuilder setOnListItemsSelected(MultiChoiceListDialogListener listener) {
            this.iMultiChoiceListDialogListener = listener;
            return this;
        }

        @Override
        public PriorityListDialog build() {
            return showAllowingStateLoss();
        }

        private PriorityListDialog create() {

            final Bundle args = new Bundle();

            final PriorityListDialog
                fragment = (PriorityListDialog) Fragment.instantiate(mContext, mClass.getName(), args);
            args.putBoolean(ARG_CANCELABLE_ON_TOUCH_OUTSIDE, mCancelableOnTouchOutside);
            args.putBoolean(ARG_USE_DARK_THEME, mUseDarkTheme);
            args.putBoolean(ARG_USE_LIGHT_THEME, mUseLightTheme);

            if (mTargetFragment != null) {
                fragment.setTargetFragment(mTargetFragment, mRequestCode);
            } else {
                args.putInt(ARG_REQUEST_CODE, mRequestCode);
            }
            fragment.setCancelable(mCancelable);

            return fragment;
        }

        public PriorityListDialog show() {
            PriorityListDialog fragment = create();
            initFragment(fragment);
            fragment.show(mFragmentManager, mTag);
            return fragment;
        }

        public PriorityListDialog showAllowingStateLoss() {
            PriorityListDialog fragment = create();
            initFragment(fragment);
            fragment.showAllowingStateLoss(mFragmentManager, mTag);
            return fragment;
        }
        private void initFragment(PriorityListDialog fragment){

            fragment.mTitle = this.title;
            fragment.mItems = this.items;
            fragment.mMode = this.mode;
            fragment.mCheckedItems = this.checkedItems;
            fragment.mCancelButtonText = this.cancelButtonText;
            fragment.mConfirmButtonText = this.confirmButtonText;
            fragment.mListDialogListener = this.iListDialogListener;
            fragment.mMultiChoiceListDialogListener = this.iMultiChoiceListDialogListener;
            fragment.mTag = this.mTag;
        }
    }
}

File: library/src/main/java/com/dino/lib/dialog/list/PriorityListDialog.kt
package com.dino.lib.dialog.list

import android.content.Context
import android.content.res.Resources
import android.os.Bundle
import android.text.TextUtils
import android.util.SparseBooleanArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AbsListView
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ListAdapter
import android.widget.ListView
import android.widget.TextView
import androidx.annotation.IntDef
import com.dino.lib.dialog.BaseDialogBuilder
import com.dino.lib.dialog.BaseDialogFragment
import com.dino.lib.dialog.PriorityMode
import com.dino.lib.dialog.PriorityMode.LOW
import com.dino.lib.dialog.R
import java.util.Arrays

/**
 * Created by dineshkumar.m on 11/03/18.
 */

class PriorityListDialog : BaseDialogFragment() {

  private var title: CharSequence? = null
  private var items: Array<CharSequence>? = null
  @ChoiceMode @get:ChoiceMode private var mode: Int = 0
  private var checkedItems: IntArray? = intArrayOf()
  private var negativeButtonText: CharSequence? = null
  private var positiveButtonText: CharSequence? = null
  private var mTag: String? = null
  private var mListDialogListener: ListDialogListener? = null
  private var mMultiChoiceListDialogListener: MultiChoiceListDialogListener? = null

  override val customTag: String?
    get() = null

  override val priority: PriorityMode = LOW


  override fun onActivityCreated(savedInstanceState: Bundle?) {
    super.onActivityCreated(savedInstanceState)
    if (arguments == null) {
      throw IllegalArgumentException(
          "use SimpleListDialogBuilder to construct this dialog"
      )
    }
  }

  private fun prepareAdapter(itemLayoutId: Int): ListAdapter {
    return object : ArrayAdapter<Any>(
        activity, itemLayoutId, R.id.pdl_text, items!!
    ) {

      /**
       * Overriding default implementation because it ignores current light/dark theme.
       */
      override fun getView(
        position: Int,
        convertView: View?,
        parent: ViewGroup
      ): View {
        var convertView = convertView
        if (convertView == null) {
          convertView = LayoutInflater.from(parent.context)
              .inflate(itemLayoutId, parent, false)
        }
        val t = convertView!!.findViewById<View>(R.id.pdl_text) as TextView
        if (t != null) {
          t.text = getItem(position) as CharSequence
        }
        return convertView
      }
    }
  }

  private fun buildMultiChoice(builder: Builder) {
    builder.setItems(prepareAdapter(R.layout.view_list_item_multichoice), checkedItems,
        AbsListView.CHOICE_MODE_MULTIPLE, AdapterView.OnItemClickListener { parent, _, _, _ ->
      val checkedPositions = (parent as ListView).checkedItemPositions
      checkedItems = asIntArray(checkedPositions)
    })
  }

  private fun buildSingleChoice(builder: Builder) {
    builder.setItems(prepareAdapter(R.layout.view_list_item_singlechoice), checkedItems,
        AbsListView.CHOICE_MODE_SINGLE, AdapterView.OnItemClickListener { parent, _, _, _ ->
      val checkedPositions = (parent as ListView).checkedItemPositions
      checkedItems = asIntArray(checkedPositions)
    })
  }

  private fun buildNormalChoice(builder: Builder) {
    builder.setItems(prepareAdapter(R.layout.view_list_item), -1,
        AdapterView.OnItemClickListener { _, _, position, id ->
          /*for (IListDialogListener listener : getSingleDialogListeners()) {
                                listener.onListItemSelected(getItems()[position], position, mRequestCode);
                            }*/
          dismiss()
          if (mListDialogListener != null) {
            mListDialogListener!!.onListItemSelected(items!![position], position, mRequestCode)
          }
        })
  }

  override fun build(builder: Builder): Builder? {
    val title = title
    if (!TextUtils.isEmpty(title)) {
      builder.setTitle(title)
    }

    if (!TextUtils.isEmpty(negativeButtonText)) {
      builder.setNegativeButton(negativeButtonText, View.OnClickListener { dismiss() })
    }

    if (mode != AbsListView.CHOICE_MODE_NONE) {
      var positiveButtonClickListener: View.OnClickListener? = null
      when (mode) {
        AbsListView.CHOICE_MODE_MULTIPLE -> positiveButtonClickListener = View.OnClickListener {
          val checkedPositions = checkedItems
          val items = items
          val checkedValues = checkedPositions?.size?.let { it1 -> arrayOfNulls<CharSequence>(it1) }
          var i = 0
          if (checkedPositions != null) {
            for (checkedPosition in checkedPositions) {
              if (checkedPosition >= 0 && checkedPosition < items!!.size) {
                checkedValues?.set(i++, items[checkedPosition])
              }
            }
          }
          dismiss()
          if (mMultiChoiceListDialogListener != null) {
            mMultiChoiceListDialogListener!!.onListItemsSelected(
                checkedValues, checkedPositions, mRequestCode
            )
          }
        }
        AbsListView.CHOICE_MODE_SINGLE -> positiveButtonClickListener = View.OnClickListener {
          var selectedPosition = -1
          val checkedPositions = checkedItems
          val items = items
          if (checkedPositions != null) {
            for (i in checkedPositions) {
              if (i >= 0 && i < items!!.size) {
                //1st valid value
                selectedPosition = i
                break
              }
            }
          }

          dismiss()
          if (selectedPosition != -1) {
            if (mListDialogListener != null) {
              mListDialogListener!!.onListItemSelected(
                  items!![selectedPosition], selectedPosition, mRequestCode
              )
            }
          } else {
            //for (ISimpleDialogCancelListener listener : getCancelListeners()) {
            //    listener.onCancelled(mRequestCode);
            //}
          }
        }
      }

      var positiveButton = positiveButtonText
      if (TextUtils.isEmpty(positiveButtonText)) {
        //we always need confirm button when CHOICE_MODE_SINGLE or CHOICE_MODE_MULTIPLE
        positiveButton = getString(android.R.string.ok)
      }
      builder.setPositiveButton(positiveButton, positiveButtonClickListener)
    }

    // prepare list and its item click listener
    val items = items
    if (items != null && items.size > 0) {
      when (mode) {
        AbsListView.CHOICE_MODE_MULTIPLE -> buildMultiChoice(builder)
        AbsListView.CHOICE_MODE_SINGLE -> buildSingleChoice(builder)
        AbsListView.CHOICE_MODE_NONE -> buildNormalChoice(builder)
      }
    }

    return builder
  }

  @IntDef(
      AbsListView.CHOICE_MODE_MULTIPLE, AbsListView.CHOICE_MODE_SINGLE,
      AbsListView.CHOICE_MODE_NONE
  )
  annotation class ChoiceMode

  class SimpleListDialogBuilder(
    context: Context,
    fragmentManager: androidx.fragment.app.FragmentManager
  ) : BaseDialogBuilder<SimpleListDialogBuilder>(
      context, fragmentManager, PriorityListDialog::class.java
  ) {

    private var title: CharSequence? = null

    private var items: Array<CharSequence>? = null

    @ChoiceMode private var mode: Int = 0
    private var checkedItems: IntArray? = null

    private var cancelButtonText: CharSequence? = null
    private var confirmButtonText: CharSequence? = null
    private var iListDialogListener: ListDialogListener? = null
    private var iMultiChoiceListDialogListener: MultiChoiceListDialogListener? = null

    private val resources: Resources
      get() = mContext.resources

    override fun self(): SimpleListDialogBuilder {
      return this
    }

    fun setTitle(title: CharSequence): SimpleListDialogBuilder {
      this.title = title
      return this
    }

    fun setTitle(titleResID: Int): SimpleListDialogBuilder {
      this.title = resources.getString(titleResID)
      return this
    }

    fun setCheckedItems(positions: IntArray): SimpleListDialogBuilder {
      this.checkedItems = positions
      return this
    }

    fun setSelectedItem(position: Int): SimpleListDialogBuilder {
      this.checkedItems = intArrayOf(position)
      return this
    }

    fun setChoiceMode(@ChoiceMode choiceMode: Int): SimpleListDialogBuilder {
      this.mode = choiceMode
      return this
    }

    fun setItems(items: Array<CharSequence>): SimpleListDialogBuilder {
      this.items = items
      return this
    }

    /*fun setItems(itemsArrayResID: Int): SimpleListDialogBuilder {
      this.items = resources.getStringArray(itemsArrayResID)
      return this
    }*/

    fun setConfirmButtonText(text: CharSequence): SimpleListDialogBuilder {
      this.confirmButtonText = text
      return this
    }

    fun setConfirmButtonText(confirmBttTextResID: Int): SimpleListDialogBuilder {
      this.confirmButtonText = resources.getString(confirmBttTextResID)
      return this
    }

    fun setCancelButtonText(text: CharSequence): SimpleListDialogBuilder {
      this.cancelButtonText = text
      return this
    }

    fun setCancelButtonText(cancelBttTextResID: Int): SimpleListDialogBuilder {
      this.cancelButtonText = resources.getString(cancelBttTextResID)
      return this
    }

    override fun setTag(tag: String): SimpleListDialogBuilder {
      this.mTag = tag
      return this
    }

    fun setOnListItemSelected(listener: ListDialogListener): SimpleListDialogBuilder {
      this.iListDialogListener = listener
      return this
    }

    fun setOnListItemsSelected(listener: MultiChoiceListDialogListener): SimpleListDialogBuilder {
      this.iMultiChoiceListDialogListener = listener
      return this
    }

    override fun build(): PriorityListDialog {
      return showAllowingStateLoss()
    }

    private fun create(): PriorityListDialog {

      val args = Bundle()

      val fragment = androidx.fragment.app.Fragment.instantiate(mContext, mClass.name, args) as PriorityListDialog
      args.putBoolean(ARG_CANCELABLE_ON_TOUCH_OUTSIDE, mCancelableOnTouchOutside)
      args.putBoolean(ARG_USE_DARK_THEME, mUseDarkTheme)
      args.putBoolean(ARG_USE_LIGHT_THEME, mUseLightTheme)

      if (mTargetFragment != null) {
        fragment.setTargetFragment(mTargetFragment, mRequestCode)
      } else {
        args.putInt(ARG_REQUEST_CODE, mRequestCode)
      }
      fragment.isCancelable = mCancelable

      return fragment
    }

    fun show(): PriorityListDialog {
      val fragment = create()
      initFragment(fragment)
      fragment.show(mFragmentManager, mTag)
      return fragment
    }

    private fun showAllowingStateLoss(): PriorityListDialog {
      val fragment = create()
      initFragment(fragment)
      fragment.showAllowingStateLoss(mFragmentManager, mTag)
      return fragment
    }

    private fun initFragment(fragment: PriorityListDialog) {

      fragment.title = this.title
      fragment.items = this.items
      fragment.mode = this.mode
      fragment.checkedItems = this.checkedItems
      fragment.negativeButtonText = this.cancelButtonText
      fragment.positiveButtonText = this.confirmButtonText
      fragment.mListDialogListener = this.iListDialogListener
      fragment.mMultiChoiceListDialogListener = this.iMultiChoiceListDialogListener
      fragment.mTag = this.mTag
    }
  }

  companion object {

    fun createBuilder(
      context: Context,
      fragmentManager: androidx.fragment.app.FragmentManager
    ): SimpleListDialogBuilder {
      return SimpleListDialogBuilder(context, fragmentManager)
    }

    private fun asIntArray(checkedItems: SparseBooleanArray): IntArray {
      var checked = 0
      // compute number of items
      for (i in 0 until checkedItems.size()) {
        val key = checkedItems.keyAt(i)
        if (checkedItems.get(key)) {
          ++checked
        }
      }

      val array = IntArray(checked)
      //add indexes that are checked
      var i = 0
      var j = 0
      while (i < checkedItems.size()) {
        val key = checkedItems.keyAt(i)
        if (checkedItems.get(key)) {
          array[j++] = key
        }
        i++
      }
      Arrays.sort(array)
      return array
    }
  }
}
